c++/src/kj/mutex.c++

```diff
@@ -58,7 +59,100 @@
 #include <windows.h>
 #endif
 
+#if KJ_TRACK_LOCK_BLOCKING
+#include <pthread.h>
+#endif
+
 namespace kj {
+
+#if KJ_TRACK_LOCK_BLOCKING
+static pthread_key_t waitingTypeKey;
+static pthread_key_t waitingOriginKey;
+static pthread_key_t waitingObjectKey; // Where the pointer to the mutex/once is stored.
+static pthread_key_t waitingWaiterKey; // Where the pointer to the Mutex::Waiter is stored.
+
+static bool initializedTrackLockBlockingTls = [] {
+  KJ_REQUIRE(0 == pthread_key_create(&waitingTypeKey, nullptr));
+  KJ_REQUIRE(0 == pthread_key_create(&waitingOriginKey, nullptr));
+  KJ_REQUIRE(0 == pthread_key_create(&waitingObjectKey, nullptr));
+  KJ_REQUIRE(0 == pthread_key_create(&waitingWaiterKey, nullptr));
+  return true;
+}();
+
+enum class BlockState : uintptr_t {
+  Idle,
+  Dirty,
+  MutexAcquisition,
+  CondVarFulfillment,
+  OnceInit,
+};
+
+Maybe<BlockOnReason> blockedReason() noexcept {
+  auto state = static_cast<BlockState>(reinterpret_cast<uintptr_t>(
+      pthread_getspecific(waitingTypeKey)));
+
+  switch (state) {
+    case BlockState::Idle:
+    case BlockState::Dirty:
+      return nullptr;
+    case BlockState::MutexAcquisition: {
+      auto origin = reinterpret_cast<const SourceLocation*>(pthread_getspecific(waitingOriginKey));
+      KJ_DASSERT(origin != nullptr);
+      auto m = reinterpret_cast<const _::Mutex*>(pthread_getspecific(waitingObjectKey));
+      return BlockOnReason{BlockOnMutexAcquisition{.mutex = m, .origin = origin}};
+    }
+    case BlockState::CondVarFulfillment: {
+      auto origin = reinterpret_cast<const SourceLocation*>(pthread_getspecific(waitingOriginKey));
+      KJ_DASSERT(origin != nullptr);
+      auto m = reinterpret_cast<const _::Mutex*>(pthread_getspecific(waitingObjectKey));
+      auto w = pthread_getspecific(waitingWaiterKey);
+      return BlockOnReason{BlockOnCondVarWait{.mutex = m, .waiter = w, .origin = origin}};
+    }
+    case BlockState::OnceInit: {
+      auto origin = reinterpret_cast<const SourceLocation*>(pthread_getspecific(waitingOriginKey));
+      KJ_DASSERT(origin != nullptr);
+      auto o = reinterpret_cast<const _::Once*>(pthread_getspecific(waitingObjectKey));
+      return BlockOnReason{BlockOnOnceInit{.once = o, .origin = origin}};
+    }
+  }
+}
+
+static void setCurrentThreadIsWaitingFor(const _::Mutex* mutex, const SourceLocation* origin) {
+  pthread_setspecific(waitingTypeKey, reinterpret_cast<void *>(BlockState::Dirty));
```

Comment:
You mention the importance of async signal safety in the commit description, but setting four variables one by one doesn't seem very signal safe. I guess that's the purpose of the "dirty" state here, so the signal handler can see that this was currently being modified.

Another solution would be to have a single thread-local pointer which points to a struct allocated on the stack. That way, updating the pointer is atomic. It does require that this info is always scoped to a stack frame, but `origin` already is so-scoped, so that seems fine.

On another note, it would be nice to use RAII to set and unset these values, to avoid any possibility of forgetting to unset the state on the way out.

Reply:
Yeah, what the hell was I thinking. I think at some point I had realized this & promptly forgot it?

I don't know that I can use RAII since these lifetimes aren't strictly tied to lexical scope (you could, but then you'll be superfluously saying you're not blocked for a period of time if you got a spurious wake & failed to acquire the lock)..

Reply:
Hmm if they're not tied to a lexical scope, how do you make sure the `origin` pointer stays valid?

It looks to me like `setCurrentTheradIsWaitingFor()`/`setCurrentThreadIsNoLongerWaiting()` are confined to a scope, though...

Reply:
The lifetime of the data is contained to lexical scope but whether or not the thread is blocking isn't tied to lexical scope. A thread may wake (spuriously or otherwise), fail to acquire the lock, & go back to being blocked. If I used an RAII object to do this, then a spurious wake would introduce a race condition (that a SIGINT could conceivably hit) whereby the lock is blocked but we've cleared the metadata from the TLS variables spuriously.

I think a `KJ_DEFER` is probably a better solution here to provide the guarantee that we always unset on the way out since at that point it's safe to unset multiple times on our way out.